### PR TITLE
Add support for having an empty CountingInput/CountingSource

### PR DESCRIPTION
**Pull request by lukecwik**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by lukecwik** [comment]
R: @peihe
CC: @dhalperi

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingInput.java [hunk 1]
```diff
@@ -75,17 +75,21 @@
...
 
   /**
    * Creates a {@link BoundedCountingInput} that will produce elements
    * starting from {@code startIndex} to {@code endIndex - 1}.
+   * If {@code startIndex == endIndex}, then no elements will be produced.
    */
   public static BoundedCountingInput forSubrange(long startIndex, long endIndex) {
-    checkArgument(endIndex > startIndex, "endIndex (%s) must be greater than startIndex (%s)",
-            endIndex, startIndex);
+    checkArgument(endIndex > startIndex,
```
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingInput.java]: ">="?

and, add tests

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingSource.java [hunk 2]
```diff
@@ -78,17 +78,21 @@
...
+    checkArgument(numElements >= 0,
+        "numElements (%s) must be greater than or equal to 0",
+        numElements);
     return new BoundedCountingSource(0, numElements);
   }
 
   /**
    * Creates a {@link BoundedSource} that will produce elements
-   * from {@code startIndex} to {@code endIndex - 1}.
+   * starting from {@code startIndex} to {@code endIndex - 1}.
```
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingSource.java]: what about:
"starting from {@code startIndex} (inclusive) to {@code endIndex} (exclusive)."

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingInput.java [hunk 3]
```diff
@@ -75,17 +75,21 @@
...
+    checkArgument(numElements >= 0,
+        "numElements (%s) must be greater than or equal to 0",
+        numElements);
     return new BoundedCountingInput(numElements);
   }
 
   /**
    * Creates a {@link BoundedCountingInput} that will produce elements
    * starting from {@code startIndex} to {@code endIndex - 1}.
+   * If {@code startIndex == endIndex}, then no elements will be produced.
```
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingInput.java]: what about:
"starting from {@code startIndex} (inclusive) to {@code endIndex} (exclusive)."

Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingInput.java]: That still may leave some ambiguity for when the values are equal.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingInput.java (hunk 1 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingInput.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingSource.java (hunk 2 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingSource.java]: Done as above

**Comment by peihe** [comment]
LGTM

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/CountingInputTest.java [hunk 4]
```diff
@@ -71,14 +71,35 @@ public static void addCountingAsserts(PCollection<Long> input, long start, long
   public void testBoundedInput() {
     Pipeline p = TestPipeline.create();
     long numElements = 1000;
-    PCollection<Long> input = p.apply(CountingInput.upTo(numElements));
+    PCollection<Long> input = p.apply(Read.from(CountingSource.upTo(numElements)));
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/CountingInputTest.java]: revert? should be testing `CountingInput`

**Comment by dhalperi** [comment]
Also, JIRA issue for this change?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/CountingInputTest.java (hunk 4 above)
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/CountingInputTest.java]: I fixed this in https://github.com/apache/incubator-beam/commit/5169e492578a3759e20b50e36ace61bc86636ad2

**Pull request closed without merging**